### PR TITLE
Remove Greaselion dependency from Rewards service (v2)

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/34237

### Overview

Currently, Brave Creator "pages" are detected on some social media sites (on desktop) by using a combination of the persistent Brave extension and Greaselion extension content scripts. For performance reasons, we must migrate away from using a persistent background page, and for maintainability reasons (and to allow the feature to work cross-platform in the future) we must migrate away from Greaselion.

With this change, the Rewards service no longer uses Greaselion. Instead, scripts are injected by the `RewardsTabHelper` into the "Brave internal" isolated world (using a new `CreatorDetectionScriptInjector` class). The scripts are individually compiled and bundled for each site from TypeScript using our current TypeScript bundling process. They are executed in the "Brave internal" isolated world. In the future, the scripts can be moved out of `brave-core` and delivered via `ComponentUpdater`.

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/brave_rewards/rewards_tab_helper.cc [hunk 1]
```diff
@@ -102,20 +104,26 @@ void RewardsTabHelper::DidFinishLoad(
   rewards_service_->OnLoad(tab_id_, validated_url);
 }
 
-void RewardsTabHelper::DidFinishNavigation(content::NavigationHandle* handle) {
-  if (!handle->IsInMainFrame() || !handle->HasCommitted() ||
-      handle->IsDownload()) {
+void RewardsTabHelper::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (!navigation_handle->HasCommitted() ||
+      !navigation_handle->IsInPrimaryMainFrame() || !rewards_service_) {
```
Comment (diracdeltas) [inline browser/brave_rewards/rewards_tab_helper.cc:110]: no need to check IsDownload() anymore? is that covered by IsInPrimaryMainFrame() ?

File: browser/brave_rewards/rewards_tab_helper.cc [hunk 2]
```diff
@@ -102,20 +104,26 @@ void RewardsTabHelper::DidFinishLoad(
...
   }
 
-  auto id = GetPublisherIdFromURL(GetWebContents().GetLastCommittedURL());
-  SetPublisherIdForTab(id ? *id : "");
-
-  MaybeSavePublisherInfo();
+  auto* rfh = navigation_handle->GetRenderFrameHost();
 
-  if (rewards_service_) {
+  if (!navigation_handle->IsSameDocument()) {
```
Comment (diracdeltas) [inline browser/brave_rewards/rewards_tab_helper.cc:116]: is it worth checking here to make sure that the URL is http(s)? otherwise it's not eliglble for rewards anyway

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 1 above)
Comment (diracdeltas) [reply browser/brave_rewards/rewards_tab_helper.cc:110]: also there is no rewards_service_ for incognito tabs, is that correct?

File: browser/brave_rewards/test/creator_detection_browsertest.cc [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+  AddBlankTabAndShow(browser());
+
+  base::RunLoop run_loop;
+  base::OneShotTimer timeout;
+  timeout.Start(FROM_HERE, TestTimeouts::action_timeout(),
+                run_loop.QuitClosure());
+
+  NavigateTo("www.youtube.com", "/@testuser");
+  EXPECT_EQ(GetRewardsTabHelper().GetPublisherIdForTab(), "");
+}
```
Comment (diracdeltas) [inline browser/brave_rewards/test/creator_detection_browsertest.cc:329]: worth adding a test that nothing is processed if rewards is enabled but in an incognito tab?

File: renderer/brave_content_renderer_client.cc [hunk 4]
```diff
@@ -63,6 +65,27 @@
...
+  static bool initialized = false;
+  if (initialized) {
+    return;
+  }
+
+  initialized = true;
+
+  // Set an empty CSP so that the main world's CSP is not used in the isolated
+  // world.
+  constexpr char kContentSecurityPolicy[] = "";
```
Comment (diracdeltas) [inline renderer/brave_content_renderer_client.cc:79]: i thought there was a default CSP for isolated worlds? if not we should set it to be something more restrictive than empty.

File: renderer/brave_content_renderer_client.cc [hunk 5]
```diff
@@ -63,6 +65,27 @@
...
+
+  blink::WebIsolatedWorldInfo info;
+  info.security_origin =
+      blink::WebSecurityOrigin::Create(GURL("chrome://brave_internal"));
```
Comment (diracdeltas) [inline renderer/brave_content_renderer_client.cc:83]: why do we have to set a security origin?

File: renderer/brave_content_renderer_client.cc (hunk 4 above)
Comment (diracdeltas) [reply renderer/brave_content_renderer_client.cc:79]: if we need to set it we should probably start with

```
        "default-src 'none'; frame-ancestors 'none'; base-uri 'none';"                                              
        "connect-src 'self';"                                                   
        "script-src 'self';"                      
```

**Review by emerick — APPROVED** [review]

File: renderer/brave_content_renderer_client.cc (hunk 4 above)
Comment (atuchin-m) [reply renderer/brave_content_renderer_client.cc:79]: @diracdeltas let's make this change in a dedicated PR?
Some features could violate this policy now => some adjustments are needed.

cc @antonok-edm

File: renderer/brave_content_renderer_client.cc (hunk 5 above)
Comment (diracdeltas) [reply renderer/brave_content_renderer_client.cc:83]: actually i see in the docs that this is needed if the CSP is set, but it's not set in Brave Translate which also sets a CSP? https://github.com/brave/brave-core/blob/ea08eba2f949b38917742c32707ebf30aa859395/chromium_src/components/translate/content/renderer/isolated_world_util.cc#L17

File: renderer/brave_content_renderer_client.cc (hunk 4 above)
Comment (diracdeltas) [reply renderer/brave_content_renderer_client.cc:79]: i just remembered we use the same ISOLATED_WORLD_ID_BRAVE_INTERNAL for a bunch of other stuff (playlist, speedreader, ads, cosmetic filtering). is this change going to affect those features as well?

also it's not clear to me how this code interacts with https://github.com/brave/brave-core/blob/ea08eba2f949b38917742c32707ebf30aa859395/components/cosmetic_filters/renderer/cosmetic_filters_js_render_frame_observer.cc#L48 since AFAICT this is using the same world id

Comment (atuchin-m) [reply renderer/brave_content_renderer_client.cc:79]: yep, if we change CSP here it will affect all there features.
That's why I suggest to split out it from the current PR.

File: renderer/brave_content_renderer_client.cc (hunk 5 above)
Comment (atuchin-m) [reply renderer/brave_content_renderer_client.cc:83]: It's a method to adjust the upstream policy. It reuses the security origin is set here:
https://source.chromium.org/chromium/chromium/src/+/main:components/translate/content/renderer/isolated_world_util.cc;l=34-35

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 2 above)
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc:121]: Yes - I've added the check to `IsMediaPlatformURL`, which is used by `CreatorDetectionAgent` when determining whether to inject a script.

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 1 above)
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc]: The documentation for `HasCommitted` suggests that it will not be true for downloads. However, just to be conservative I've added back the download check.

Yes - there's no RewardsService for incognito tabs.

File: renderer/brave_content_renderer_client.cc (hunk 5 above)
Comment (bridiver) [reply renderer/brave_content_renderer_client.cc:83]: This seems like a good change, but maybe it should be a separate PR now since we're not doing anything on the renderer side anymore in this PR?

File: browser/brave_rewards/creator_detection_script_injector.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+#ifndef BRAVE_BROWSER_BRAVE_REWARDS_CREATOR_DETECTION_SCRIPT_INJECTOR_H_
+#define BRAVE_BROWSER_BRAVE_REWARDS_CREATOR_DETECTION_SCRIPT_INJECTOR_H_
+
+#include <optional>
+#include <string>
+#include <string_view>
+
+#include "base/functional/callback.h"
+#include "base/memory/raw_ref.h"
```
Comment (bridiver) [inline browser/brave_rewards/creator_detection_script_injector.h:14]: icwyu should be raw_ptr.h

File: browser/brave_rewards/test/creator_detection_browsertest.cc [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+)";
+
+IN_PROC_BROWSER_TEST_F(CreatorDetectionBrowserTest, YouTubeDetection) {
+  SetRewardsEnabled();
+  SetResponseHTML(kYouTubeHTML);
+
+  AddBlankTabAndShow(browser());
+  base::test::TestFuture<std::string> future;
+  TabHelperObserver observer(GetRewardsTabHelper(), future.GetCallback());
+  NavigateTo("www.youtube.com", "/@testuser");
```
Comment (bridiver) [inline browser/brave_rewards/test/creator_detection_browsertest.cc:299]: Seems like we should do a "same page" navigation as part of this test since that's what YT does?

File: browser/brave_rewards/rewards_tab_helper.cc [hunk 8]
```diff
@@ -192,13 +205,43 @@ void RewardsTabHelper::MaybeSavePublisherInfo() {
...
+    visit->domain = result->id;
+    visit->name = result->name;
+    visit->path = "";
+    visit->url = result->url;
+    visit->favicon_url = result->image_url;
+    if (auto platform = GetMediaPlatformFromPublisherId(result->id)) {
+      visit->provider = *platform;
+    }
+
+    rewards_service_->GetPublisherActivityFromUrl(visit->Clone());
```
Comment (bridiver) [inline browser/brave_rewards/rewards_tab_helper.cc:239]: this should really be encapsulated in something like `UpdateCreatorForTab`. I have no idea what is going on in these three calls and this also makes it hard to test vs unit testing a single method on RewardsService. We really shouldn't be exposing these kinds of low level calls in the service.

File: browser/brave_rewards/creator_detection_script_injector.cc [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+void CreatorDetectionScriptInjector::DetectCreator(
+    content::RenderFrameHost* rfh,
+    DetectCreatorCallback callback) {
+  if (!rfh || rfh != injector_host_.get() || !injector_.is_bound()) {
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+
+  // Do not call the detection routine if it has already been run for this URL.
```
Comment (bridiver) [inline browser/brave_rewards/creator_detection_script_injector.cc:133]: Seems like this is something that the tab helper should be responsible for?

File: browser/brave_rewards/creator_detection_script_injector.h [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+ private:
+  using ExecuteScriptCallback = base::OnceCallback<void(base::Value)>;
+
+  void ExecuteScript(std::string_view script, ExecuteScriptCallback callback);
+
+  void OnCreatorDetected(DetectCreatorCallback callback, base::Value value);
+
+  mojo::AssociatedRemote<script_injector::mojom::ScriptInjector> injector_;
+  raw_ptr<content::RenderFrameHost> injector_host_ = nullptr;
```
Comment (bridiver) [inline browser/brave_rewards/creator_detection_script_injector.h:74]: It's not safe to hold a reference to a RenderFrameHost like this. The rfh can be destroyed by the time `DetectCreator` is called. Store `GlobalRenderFrameHostId` instead and then check for null when calling RenderFramehost::FromID

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 8 above)
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc:239]: side note: why is this called `GetPublisherActivityFromUrl` if it doesn't actually return anything? Also it doesn't take a URL as a param which is confusing. Separately from this PR this method name should probably change.

File: browser/brave_rewards/test/creator_detection_browsertest.cc [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+  }
+
+  void SetRewardsEnabled() {
+    auto* prefs = browser()->profile()->GetPrefs();
+    prefs->SetBoolean(brave_rewards::prefs::kEnabled, true);
+  }
+
+  void NavigateTo(std::string_view host, std::string_view path) {
+    auto url = embedded_https_test_server().GetURL(host, path);
+    ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), url));
```
Comment (bridiver) [inline browser/brave_rewards/test/creator_detection_browsertest.cc:83]: This pattern of assertions within helper methods obscures the source of the failure. It's better to return a bool and `ASSERT(NavigateTo(...))`

File: browser/brave_rewards/test/creator_detection_browsertest.cc [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+    return response;
+  }
+
+  base::test::ScopedFeatureList scoped_feature_list_;
+  RequestCallback request_callback_;
+  std::string response_html_;
+};
+
+IN_PROC_BROWSER_TEST_F(CreatorDetectionBrowserTest, GithubDetection) {
+  SetRewardsEnabled();
```
Comment (bridiver) [inline browser/brave_rewards/test/creator_detection_browsertest.cc:132]: We also need to check that it doesn't do anything when rewards is not enabled. Also if you have something that needs to run at the start of every test it should go in SetUp/SetUpOnMainThread

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 8 above)
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc:239]: oh, I guess an overloaded version of `GetPublisherActivityFromUrl` was added in this PR so the name should at least change to `GetPublisherActivityFromVisitData` and then change `GetPublisherActivityFrom` for both in a separate PR

File: browser/brave_rewards/test/creator_detection_browsertest.cc (hunk 11 above)
Comment (atuchin-m) [reply browser/brave_rewards/test/creator_detection_browsertest.cc:83]: Also, IMHO it's okay to use `CHECK_EQ` if it's related to the correctness of the text fixture (not for the subsystem we're going to verify).
`ASSERT_TRUE(SomeHelperFunction(..))` makes the text harder to read.

File: browser/brave_rewards/test/creator_detection_browsertest.cc (hunk 7 above)
Comment (atuchin-m) [reply browser/brave_rewards/test/creator_detection_browsertest.cc:299]: Sound like we need to cover the both cases.

File: browser/brave_rewards/test/creator_detection_browsertest.cc [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+
+IN_PROC_BROWSER_TEST_F(CreatorDetectionBrowserTest, InvalidHost) {
+  SetRewardsEnabled();
+  SetResponseHTML(kYouTubeHTML);
+  AddBlankTabAndShow(browser());
+
+  base::RunLoop run_loop;
+  base::OneShotTimer timeout;
+  timeout.Start(FROM_HERE, TestTimeouts::action_timeout(),
+                run_loop.QuitClosure());
```
Comment (atuchin-m) [inline browser/brave_rewards/test/creator_detection_browsertest.cc:312]: We suppose to  call `run_loop.Run()`, don't we?

nit: maybe we can make a helper method `WaitForTimeout(..)`?

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 14]
```diff
@@ -1478,17 +1429,32 @@ void RewardsServiceImpl::GetBalanceReport(
                      std::move(callback)));
 }
 
+void RewardsServiceImpl::GetPublisherActivityFromUrl(
+    mojom::VisitDataPtr visit_data) {
+  if (!Connected()) {
+    return;
+  }
+  uint32_t tab_id = visit_data->tab_id;
+  engine_->GetPublisherActivityFromUrl(tab_id, std::move(visit_data), "");
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc:1438]: I'm confused, below we're passing `windowId` to `GetPublisherActivityFromUrl`, but we're passing tabId here? Why would we ever be dealing with a window id?

File: browser/brave_rewards/creator_detection_script_injector.h (hunk 6 above)
Comment (zenparsing) [reply browser/brave_rewards/creator_detection_script_injector.h]: Fixed (`raw_ptr` is actually no longer required).

File: browser/brave_rewards/rewards_tab_helper.cc [hunk 15]
```diff
@@ -192,13 +205,43 @@ void RewardsTabHelper::MaybeSavePublisherInfo() {
...
       tab_id_.id(), GetWebContents().GetLastCommittedURL().spec(), "", "");
 }
 
+void RewardsTabHelper::OnCreatorDetected(
+    std::optional<CreatorDetectionScriptInjector::Result> result) {
+  if (!rewards_service_ || !result) {
+    return;
+  }
+
+  SetPublisherIdForTab(result->id);
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc:228]: Thanks for catching this! I've updated this callback to check that the URL still matches.

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 8 above)
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc]: Yes - these methods are pretty ancient and the names reflect that. I've updated the new method name as suggested, and I'll add an issue to fix the names up more generally.

File: browser/brave_rewards/creator_detection_script_injector.cc (hunk 9 above)
Comment (zenparsing) [reply browser/brave_rewards/creator_detection_script_injector.cc]: Moved this check to `RewardsTabHelper`.

File: browser/brave_rewards/test/creator_detection_browsertest.cc (hunk 11 above)
Comment (zenparsing) [reply browser/brave_rewards/test/creator_detection_browsertest.cc]: Fixed.

File: browser/brave_rewards/test/creator_detection_browsertest.cc (hunk 13 above)
Comment (zenparsing) [reply browser/brave_rewards/test/creator_detection_browsertest.cc]: Yes - thanks.

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 14 above)
Comment (zenparsing) [reply components/brave_rewards/browser/rewards_service_impl.cc:1436]: This is a legacy function - the `windowId` parameter name might event predate `brave-core`. I've updated the parameter name to be `tab_id` to reduce confusion.

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc:228]: do we want to completely ignore the result in this case and not send anything to the rewards service or do we just not want to update the publisher? Won't we lose the publisher data below
```
// When a creator has been detected for the current tab, we must send the
    // creator data to the utility process so that the "publisher_info" database
    // table can be populated
```

File: renderer/brave_content_renderer_client.cc (hunk 5 above)
Comment (bridiver) [reply renderer/brave_content_renderer_client.cc:83]: I believe the isolated world is automatically created the first time you attempt to execute JS. We have lots of other JS execution in isolated worlds outside of cosmetic filtering. Are you sure it doesn't work without this?

File: browser/brave_rewards/creator_detection_script_injector.cc [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    return;
+  }
+
+  injector_host_token_ = rfh->GetGlobalFrameToken();
+  rfh->GetRemoteAssociatedInterfaces()->GetInterface(&injector_);
+
+  bool verbose_logging =
+      base::FeatureList::IsEnabled(features::kVerboseLoggingFeature);
+
+  // Add a `braveRewards` global variable to the isolated world.
```
Comment (bridiver) [inline browser/brave_rewards/creator_detection_script_injector.cc:118]: I'm a bit confused about why we're injecting the script separately from running the function. Why don't we just wrap the script with a call to the function and return the result? This seems to complicate the process and it's not clear why we need to do this in two steps?

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 8 above)
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc]: what about my comment about wrapping these three method calls in a single call to the service? If multiple actions need to be taken in response to something it's best for the service to handle that, not the caller. Although this is not a user action, it should still follow the same principle of having a single call to the service for a given event that occurs outside the component

File: browser/brave_rewards/creator_detection_script_injector.cc (hunk 16 above)
Comment (bridiver) [reply browser/brave_rewards/creator_detection_script_injector.cc:118]: Sorry I missed this on the first pass

File: renderer/brave_content_renderer_client.cc (hunk 5 above)
Comment (zenparsing) [reply renderer/brave_content_renderer_client.cc:83]: For the twitter script, we're using the same technique as cosmetic filters (inserting a script tag in order to execute a small bit of JS in the main world), which requires the CSP customization.

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc:228]: I think that if the URL changes before the detection callback is run, then I think it's OK to treat it as if the creator info could not be detected for that previous page load.

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 8 above)
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc]: IMO, the `RewardsService` API isn't great, but refactoring `RewardsService` is more of a separate concern.

File: browser/brave_rewards/creator_detection_script_injector.cc (hunk 16 above)
Comment (zenparsing) [reply browser/brave_rewards/creator_detection_script_injector.cc:118]: The bundled scripts aren't small, and it would be more efficient to inject the bundle once and then call it as needed (e.g. for subsequent same-doc navigations). I'm not sure if it would be an overall improvement to execute the bundle each time.

File: renderer/brave_content_renderer_client.cc (hunk 5 above)
Comment (zenparsing) [reply renderer/brave_content_renderer_client.cc:83]: @diracdeltas I suppose that we could introduce a new world ID and unconditionally initialize it (i.e. initialize it for all renderers) similar to what we're doing here with `ISOLATED_WORLD_ID_BRAVE_INTERNAL`.

The current state, though, is that everything using `ISOLATED_WORLD_ID_BRAVE_INTERNAL` is using the custom CSP (as long as the cosmetic filters feature is not disabled, that is). Seems a bit out of scope to rework isolated worlds here.

I could also just back out this change and let the Twitter detection fail if cosmetic filters is disabled. Would that be preferrable?

Comment (diracdeltas) [reply renderer/brave_content_renderer_client.cc:83]: > I could also just back out this change and let the Twitter detection fail if cosmetic filters is disabled. Would that be preferable?

yeah i think we should do that for this PR and have a separate issue/PR for moving cosmetic filters to its own world and twitter detection to its own world

Comment (zenparsing) [reply renderer/brave_content_renderer_client.cc]: I've removed the changes on the renderer side.

**Review by diracdeltas — APPROVED** [review]

File: browser/brave_rewards/test/creator_detection_browsertest.cc (hunk 12 above)
Comment (bridiver) [reply browser/brave_rewards/test/creator_detection_browsertest.cc:139]: it's pretty easy to miss, I think it would be better to check enabled/disabled for each test so `SetRewardsEnabled(true) .... SetRewardsEnabled(false) ...`? We should also test edge cases like navigating before we get the response from the script.

File: browser/brave_rewards/creator_detection_script_injector.cc [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+      blink::mojom::UserActivationOption::kDoNotActivate,
+      blink::mojom::PromiseResultOption::kAwait, std::move(callback));
+}
+
+void CreatorDetectionScriptInjector::OnCreatorDetected(
+    DetectCreatorCallback callback,
+    base::Value value) {
+  Result result;
+  if (auto* dict = value.GetIfDict()) {
+    if (auto* id = dict->FindString("id")) {
```
Comment (bridiver) [inline browser/brave_rewards/creator_detection_script_injector.cc:160]: is it ok for some/all of these to be null? Does something else validate it later?

File: browser/brave_rewards/test/creator_detection_browsertest.cc [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+  WaitForTimeout();
+  EXPECT_EQ(GetRewardsTabHelper().GetPublisherIdForTab(), "");
+}
+
+IN_PROC_BROWSER_TEST_F(CreatorDetectionBrowserTest, RewardsDisabled) {
+  SetResponseHTML(kYouTubeHTML);
+  AddBlankTabAndShow(browser());
+  ASSERT_TRUE(NavigateTo("www.youtube.com", "/@testuser"));
+  WaitForTimeout();
+  EXPECT_EQ(GetRewardsTabHelper().GetPublisherIdForTab(), "");
```
Comment (bridiver) [inline browser/brave_rewards/test/creator_detection_browsertest.cc:345]: other things happen in RewardsTabHelper like `MaybeSavePublisherInfo` that we don't appear to be testing anywhere.

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 8 above)
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc]: I'm not suggesting refactoring the entire class, but we can certainly simplify this specific case, no? Just because other code is problematic doesn't mean we should keep doing the same thing when we could at least wrap these calls. It's only adding a couple of methods to RewardsService to move the details inside of it

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc:228]: I feel like we should do something more robust here than checking the url like having a request id that we check.

File: browser/brave_rewards/rewards_tab_helper.cc [hunk 19]
```diff
@@ -192,13 +210,45 @@ void RewardsTabHelper::MaybeSavePublisherInfo() {
...
   // the web. Previously, this was accomplished within the background script of
   // the Rewards extension.
   rewards_service_->GetPublisherActivityFromUrl(
       tab_id_.id(), GetWebContents().GetLastCommittedURL().spec(), "", "");
 }
 
+void RewardsTabHelper::OnCreatorDetected(
+    GURL url,
+    std::optional<CreatorDetectionScriptInjector::Result> result) {
+  if (!rewards_service_ || !result ||
```
Comment (bridiver) [inline browser/brave_rewards/rewards_tab_helper.cc:223]: I think this should not be possible, right? It would never trigger the script in the first place if rewards_service_ is null so this should be a `CHECK(rewards_service_)`

File: browser/brave_rewards/rewards_tab_helper.cc [hunk 20]
```diff
@@ -98,19 +100,35 @@ void RewardsTabHelper::DidFinishLoad(
...
   }
 
-  auto id = GetPublisherIdFromURL(GetWebContents().GetLastCommittedURL());
-  SetPublisherIdForTab(id ? *id : "");
+  if (!rewards_service_) {
+    return;
+  }
 
-  MaybeSavePublisherInfo();
+  auto& url = GetWebContents().GetLastCommittedURL();
```
Comment (bridiver) [inline browser/brave_rewards/rewards_tab_helper.cc:115]: don't use `GetWebContents().GetLastCommittedURL()` when NavigationHandle is available

File: browser/brave_rewards/test/creator_detection_browsertest.cc (hunk 18 above)
Comment (bridiver) [reply browser/brave_rewards/test/creator_detection_browsertest.cc:345]: sorry, I guess `MaybeSavePublisherInfo` was already in there so hopefully it's tested somewhere else, but it doesn't look like we're testing that this stuff happens https://github.com/brave/brave-core/pull/25949/files#diff-8126a7ac8b32edde8721529071011a4f05cbb82024ec55561f8af093eed86912R246

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 19 above)
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc:223]: I guess to be extra cautious maybe a DCHECK instead of CHECK for now and we can convert to a CHECK later

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc:227]: I've updated the check to use `last_detection_url_`, and I think that this is sufficient.

There may be the possibility that a user quickly visits urls `A -> B -> A`, and then we get a detection result for the first `A` and interpret it as a result for the second `A`. The detection result, however, will be valid for either visit.

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 8 above)
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc]: Filed a follow-up issue here: https://github.com/brave/brave-browser/issues/41832 to address the RewardsService AC-related API issues. It will be more clear what the API should be after the upcoming AC changes have been finalized.

File: browser/brave_rewards/test/creator_detection_browsertest.cc (hunk 12 above)
Comment (zenparsing) [reply browser/brave_rewards/test/creator_detection_browsertest.cc]: Added testing for:

* Feature disabled
* Same-doc navigation
* Navigation away from a platform page
* Non-platform site creator ID
* Storage to the `publisher_info` table

The AC/activity related behavior of `RewardsTabHelper` is currently tested by `RewardsContributionBrowserTest`.

File: browser/brave_rewards/creator_detection_script_injector.cc (hunk 17 above)
Comment (zenparsing) [reply browser/brave_rewards/creator_detection_script_injector.cc:160]: It is OK. `RewardsTabHelper` checks the "id" member to determine whether creator info was detected. Each of the other members is best-effort.

File: browser/brave_rewards/test/creator_detection_browsertest.cc (hunk 18 above)
Comment (zenparsing) [reply browser/brave_rewards/test/creator_detection_browsertest.cc:414]: Tests added for ensuring that the data is stored to the database (see cases `YouTubeDetection` and `NonPlatformSite`).

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 19 above)
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc]: Converted to a `CHECK`.

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 20 above)
Comment (zenparsing) [reply browser/brave_rewards/rewards_tab_helper.cc]: Fixed - thanks.

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc:227]: I'm not so much worried about the A -> B -> A thing, it's just not a great way to handle these kinds of things and I want to discourage people from tracking based on the url. I didn't think about it  before, but there is something already here we can use `NavigationHandle::GetGlobalRequestID`

File: browser/brave_rewards/rewards_tab_helper.cc (hunk 8 above)
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc]: fyi I made a similar comment in this PR to make things more unit testable https://github.com/brave/brave-core/pull/26082/files#r1813441440

Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc]: unit testing tab helpers is not very practical, but unit testing the service is

Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc]: I won't block on this as long as we have good test coverage for everything that gets called here, but given that this is new code being added I'm struggling with why we can't just add a new method to the service to wrap these three calls so we can unit test it.

File: browser/brave_rewards/rewards_tab_helper.cc [hunk 21]
```diff
@@ -98,19 +100,35 @@ void RewardsTabHelper::DidFinishLoad(
...
+    creator_detection_.MaybeInjectScript(
+        navigation_handle->GetRenderFrameHost());
+  }
+
+  if (url != last_detection_url_) {
+    last_detection_url_ = url;
+    creator_detection_.DetectCreator(
+        navigation_handle->GetRenderFrameHost(),
+        base::BindOnce(&RewardsTabHelper::OnCreatorDetected,
+                       base::Unretained(this), url));
```
[the remark this answers is not in the document]
Comment (bridiver) [reply browser/brave_rewards/rewards_tab_helper.cc:131]: This usage is ok, RewardsTabHelper owns creator_detection_ and they are both used on the same sequence

File: browser/brave_rewards/creator_detection_script_injector.cc [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+  auto script_source = GetDetectionScript(rfh);
+  if (!script_source) {
+    return;
+  }
+
+  injector_host_token_ = rfh->GetGlobalFrameToken();
+  rfh->GetRemoteAssociatedInterfaces()->GetInterface(&injector_);
+
+  bool verbose_logging =
+      base::FeatureList::IsEnabled(features::kVerboseLoggingFeature);
```
Comment (bridiver) [inline browser/brave_rewards/creator_detection_script_injector.cc:116]: Why are we using a feature flag to control logging? Log levels should be controlled by chromium logging methods.

**Review by bridiver — APPROVED** [review]

File: browser/brave_rewards/creator_detection_script_injector.cc (hunk 22 above)
Comment (zenparsing) [reply browser/brave_rewards/creator_detection_script_injector.cc]: This logging flag has been removed; the scripts now simply use `console.debug`.
